### ISSUE TITLE
bug: `last_used_at` not updated when "activity" is sent via API `/report-stats`

### ISSUE THREAD

**Issue by ericpaulsen** [coder/coder#11986]
customers are looking to send an activity "ping" to workspaces to keep them from auto-shutting off. we have the below API call documented, but this does not update the workspace `last_used_at` time (retrieved via `/api/v2/users/{user}/workspace/{workspace-name}`):

```shell
#!/bin/bash
# Send workspace activity as long as the job is still running

while true
do
  if pgrep -f "my_training_script.py" > /dev/null
  then
    curl -X POST "https://coder.example.com/api/v2/workspaceagents/me/report-stats" \
    -H "Coder-Session-Token: $CODER_AGENT_TOKEN" \
    -d '{
      "connection_count": 1
    }'

    # Sleep for 30 minutes (1800 seconds) if the job is running
    sleep 1800
  else
    # Sleep for 1 minute (60 seconds) if the job is not running
    sleep 60
  fi
done
```

**Comment by mattlqx** [comment]
It's very desirable for the agent token to be able to bump the activity so that user login is not required to accomplish this, as actual usage is in a crontab.

Could the `report-stats` endpoint accept the `PATCH` method and then would mean only update the keys defined in the body? So that `connections_by_proto: {"ssh": 1}` would leave the other connections untouched?

**Comment by johnstcn** [comment]
@mattlqx You don't need a human to interact with the workspace in order for that to work.

You can retrieve a valid session token for the workspace owner using `data.coder_workspace.XXX.owner_session_token` ([link](https://registry.terraform.io/providers/coder/coder/latest/docs/data-sources/workspace#owner_session_token)). Similarly, if you need the workspace ID, this is accessible via `data.coder_workspace.XXX.id`.

Does that address your use case?

**Comment by mattlqx** [comment]
That helps, I'll give it a shot.

**Comment by mattlqx** [comment]
`/api/v2/workspaces/{workspace_id}/extend` could work but would require multiple API calls to get the current stop time to not set the value lower than what is already set. It also doesn't help track activity to get that Last Used display to be more accurate.

It would be much better if `report-stats` could accept partial updates.

**Comment by johnstcn** [comment]
@mattlqx Thanks, that's good feedback. We do have some plans to rework how our workspace scheduling (see: https://github.com/coder/coder/issues/11503)

**Comment by strike** [comment]
I'm not sure if this is the right place to post, but I wanted to add it here.

We also need similar functionality for dynamically extending the workspace runtime during use. We have scenarios like:
1. Direct external connections to the workspace, without using Tailscale.
2. Running processes (by name).
3. High CPU load.
It would be great if these could be taken into account.

**Comment by johnstcn** [comment]
Would the ability to specify a positive delta to `/extend` cover this use case?

e.g.

```
curl -XPUT 'https://coder.example.com/api/v2/workspaces/{workspace_id}/extend' \ 
  -H "Coder-Session-Token: $(cat ~/.config/coderv2/session)" \
  --data-raw "{\"delta\":\"1h23m45s"}"
```

**Comment by strike** [comment]
I believe a simple endpoint that updates variable like `last_human_used_at` would be beneficial. Then, connect the shutdown timing to `last_human_used_at`.

It seems that basing shutdowns on user connections was not the best approach ([PR](https://github.com/coder/coder/pull/10808/files#diff-e71d8c645ba64bced5cde8872ac613aba0975ff69969d6cc5c5a9721195d682d)).

**Comment by bamhm182** [comment]
Just here to say that I too would like the ability to update the `last_used_at` timestamp using the Agent Token. My use case is that I have been using Coder to spin up GUI-based VMs, which use external coder_apps. The method in the docs shown in the original post would be ideal, but I can also confirm it doesn't work.

**Comment by johnstcn** [comment]
https://github.com/coder/coder/pull/12659 introduces an API endpoint `POST /api/v2/workspaces/{workspace_id}/usage` which allows directly updating the `last_used_at` timestamp of a workspace. It's been merged to main and should end up in the next release cycle.

**Comment by johnstcn** [comment]
The above was included in [v2.10.0](https://github.com/coder/coder/releases/tag/v2.10.0).
